File: solutions/exercise-09/process-application/src/main/java/io/miragon/training/adapter/outbound/db/MembershipPersistenceAdapter.java
package io.miragon.training.adapter.outbound.db;

import io.miragon.training.application.port.outbound.MembershipRepository;
import io.miragon.training.domain.Membership;
import io.miragon.training.domain.MembershipId;
import org.springframework.stereotype.Component;

import java.util.NoSuchElementException;

@Component
public class MembershipPersistenceAdapter implements MembershipRepository {

    private final MembershipJpaRepository jpaRepository;
    private final MembershipEntityMapper mapper;

    public MembershipPersistenceAdapter(MembershipJpaRepository jpaRepository, MembershipEntityMapper mapper) {
        this.jpaRepository = jpaRepository;
        this.mapper = mapper;
    }

    @Override
    public Membership find(MembershipId id) {
        var entity = jpaRepository.findById(id.value())
                .orElseThrow(() -> new NoSuchElementException("Membership not found: " + id.value()));
        return mapper.toDomain(entity);
    }

    @Override
    public Membership save(Membership membership) {
        jpaRepository.save(mapper.toEntity(membership));
        return membership;
    }
}

File: solutions/exercise-7/src/main/kotlin/io/miragon/training/adapter/outbound/db/MembershipPersistenceAdapter.kt
package io.miragon.training.adapter.outbound.db

import io.miragon.training.application.port.outbound.MembershipRepository
import io.miragon.training.domain.Membership
import io.miragon.training.domain.MembershipId
import org.springframework.stereotype.Component

@Component
class MembershipPersistenceAdapter(
    private val jpaRepository: MembershipJpaRepository,
    private val mapper: MembershipEntityMapper,
) : MembershipRepository {

    override fun find(id: MembershipId): Membership {
        val entity = jpaRepository.findById(id.value)
            .orElseThrow { NoSuchElementException("Membership not found: ${id.value}") }
        return mapper.toDomain(entity)
    }

    override fun save(membership: Membership): Membership {
        val entity = mapper.toEntity(membership)
        jpaRepository.save(entity)
        return membership
    }
}
